### PR TITLE
StroeerCore: Add new bidder

### PR DESCRIPTION
**Pull request by philipwatson**

### CODE REVIEW

**Comment by philipwatson** [comment]
https://github.com/prebid/prebid.github.io/pull/3781

File: src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        assertThat(mapper.readValue(httpRequest.getBody(), BidRequest.class)).isEqualTo(expectedBidRequest);
+        assertThat(httpRequest.getMethod()).isEqualTo(HttpMethod.POST);
+        assertThat(httpRequest.getUri()).isEqualTo(ENDPOINT_URL);
+        assertThat(httpRequest.getPayload()).isEqualTo(expectedBidRequest);
+        assertThat(httpRequest.getHeaders())
+                .extracting(Map.Entry::getKey, Map.Entry::getValue)
+                .containsOnly(
+                        tuple("Content-Type", "application/json;charset=utf-8"),
+                        tuple("Accept", "application/json"));
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java:328]: It's unit tests and you need to assert only specific test cases. No need to assert whole request. For example in `makeHttpRequestsShouldReturnExpectedRequest` assertion part can be presented like:
```
assertThat(result.getValue())
                .extracting(HttpRequest::getPayload)
                .flatExtracting(BidRequest::getImp)
                .extracting(Imp::getTagid)
                .containsExactly("192848", "abc");
```

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java (hunk 1 above)
Comment (philipwatson) [reply src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java]: Made changes here. Is everything ok now?

**Comment by philipwatson** [comment]
In case it's missed in the hidden conversations above. Regarding the redundant uses of the JsonProperty annotation on `StroeerCoreBid`:
> I had to keep @JsonProperty("bidId") otherwise it sets bidId to null on the decode. This is strange.
I'm AFK for most of today. If this is unacceptable I can try to fix it tomorrow.

File: src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBid.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    @JsonProperty("cpm")
+    BigDecimal cpm;
+    @JsonProperty("width")
+    int width;
+    @JsonProperty("height")
+    int height;
+    @JsonProperty("ad")
+    String adMarkup;
+    @JsonProperty("crid")
+    String creativeId;
```
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBid.java]: That's ok. In our project `ObjectMapper` configured to serialize/deserialize properties using snake case. `@JsonProperty` overrides this approach.
Java property | JSON property
--- | ---
cpm | cpm
bidId | bid_id

File: src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBidResponse.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.prebid.server.bidder.stroeercore.model;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+
+import java.util.List;
+
+@AllArgsConstructor(staticName = "of")
+@Value
```
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBidResponse.java]: I think you accidentally missed it :)

File: src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+
+    private Imp createEmptySlotIdImpFrom(final Imp validImp) {
+        return validImp.toBuilder()
+                .ext(mapper.valueToTree(ExtPrebid.of(null, ExtImpStroeerCore.of(" ")))).build();
+    }
+
+    private Imp createImpWithFloorFrom(final Imp imp, final String bidFloorCurrency, final BigDecimal bidFloor) {
+        return imp.toBuilder().bidfloorcur(bidFloorCurrency).bidfloor(bidFloor).build();
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java:391]: I think these are superfluous methods and they only complicate the tests.

What about this approach?
```
    private Imp createImp(UnaryOperator<Imp.ImpBuilder> impCustomizer) {
        return impCustomizer
                .apply(Imp.builder().banner(Banner.builder().build()))
                .build();
    }

    private Imp createImp(String slotId, UnaryOperator<Imp.ImpBuilder> impCustomizer) {
        final ObjectNode impExtNode = mapper.valueToTree(ExtPrebid.of(null, ExtImpStroeerCore.of(slotId)));

        final UnaryOperator<Imp.ImpBuilder> addImpExt = imp -> imp.ext(impExtNode);
        return createImp(addImpExt.andThen(impCustomizer)::apply);
    }
```

Than we can easily rewrite (as example) `makeHttpRequestsShouldIgnoreInvalidImpressions` `given` part:
```
        final BidRequest bidRequest = createBidRequest(
                createImp(imp -> imp.ext(mapper.valueToTree(ExtPrebid.of(null, mapper.createArrayNode())))),
                createImp("b", identity()),
                createImp("   ", identity()),
                createImp(imp -> imp.banner(null)),
                createImp("e", imp -> imp.bidfloor(BigDecimal.ONE).bidfloorcur("GPB")),
                createImp("f", identity()));
```

File: src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+    private BidRequest createBidRequest(final String... slotIds) {
+        final ArrayList<Imp> imps = new ArrayList<>();
+
+        long impId = 1;
+        for (String slotId : slotIds) {
+            imps.add(createImp(String.valueOf(impId++), slotId));
+        }
+
+        return createBidRequest(imps.toArray(new Imp[]{}));
+    }
```
Comment (CTMBNara) [inline src/test/java/org/prebid/server/bidder/stroeercore/StroeerCoreBidderTest.java:359]: I think it's redundant method.

**Review by CTMBNara — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBid.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (philipwatson) [reply src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBid.java]: Ah, I see! Cool :)

File: src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBidResponse.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipwatson) [reply src/main/java/org/prebid/server/bidder/stroeercore/model/StroeercoreBidResponse.java]: Oh, and I missed it again! I'll put in a separate commit.

**Review by CTMBNara — APPROVED** [review]

**Comment by philipwatson** [comment]
Great! Thank you for your time and expertise, @CTMBNara :)

**Pull request merged**